bgpd/bgp_attr.c

```diff
@@ -2681,6 +2656,40 @@ bgp_attr_parse_ret_t bgp_attr_parse(struct peer *peer, struct attr *attr,
 		return BGP_ATTR_PARSE_ERROR;
 	}
 
+	/*
+	 * RFC4271: If the NEXT_HOP attribute field is syntactically incorrect,
+	 * then the Error Subcode MUST be set to Invalid NEXT_HOP Attribute.
+	 * This is implemented below and will result in a NOTIFICATION. If the
+	 * NEXT_HOP attribute is semantically incorrect, the error SHOULD be
+	 * logged, and the route SHOULD be ignored. In this case, a NOTIFICATION
+	 * message SHOULD NOT be sent. This is implemented elsewhere.
+	 *
+	 * RFC4760: An UPDATE message that carries no NLRI, other than the one
+	 * encoded in the MP_REACH_NLRI attribute, SHOULD NOT carry the NEXT_HOP
+	 * attribute. If such a message contains the NEXT_HOP attribute, the BGP
+	 * speaker that receives the message SHOULD ignore this attribute.
+	 */
+	if (CHECK_FLAG(attr->flag, ATTR_FLAG_BIT(BGP_ATTR_NEXT_HOP))
+	    && !CHECK_FLAG(attr->flag, ATTR_FLAG_BIT(BGP_ATTR_MP_REACH_NLRI))) {
+
+		in_addr_t nexthop_h;
+
+		nexthop_h = ntohl(attr->nexthop.s_addr);
+		if ((IPV4_NET0(nexthop_h) || IPV4_NET127(nexthop_h)
+		     || IPV4_CLASS_DE(nexthop_h))
+		    && !BGP_DEBUG(allow_martians, ALLOW_MARTIANS)) {
+			char buf[INET_ADDRSTRLEN];
+
+			inet_ntop(AF_INET, &attr->nexthop.s_addr, buf,
+				  INET_ADDRSTRLEN);
+			flog_err(EC_BGP_ATTR_MARTIAN_NH, "Martian nexthop %s",
+				 buf);
+			bgp_notify_send(peer, BGP_NOTIFY_UPDATE_ERR,
+					BGP_NOTIFY_UPDATE_INVAL_NEXT_HOP);
+			return BGP_ATTR_PARSE_ERROR;
+		}
+	}
+
```

Comment:
@nikos-github - the change looks good to me, but have you checked that leaving the ATTR_NEXT_HOP attr flag on does not lead to any problems elsewhere in the code?

Reply:
Yes I checked. Once we return BGP_ATTR_PARSE_ERROR, we then return BGP_Stop and the attribute itself falls out of scope. Subsequent calls to bgp_update_receive always memset the local attribute variable. Also the nexthop field is an attribute member not a pointer from the attribute.

Reply:
Hi @nikos-github, I didn't mean the case where the connection would get reset but the new scenario arising out of your change where the next hop is simply ignored if MP nexthop is also present. In this case, the attr flag will have ATTR_NEXT_HOP set even though what the code should be using is the MP nexthop. My hunch/recollection is that this should be fine, but calling attention to it so you also can evaluate.

Reply:
Let me check on a couple of things. @srimohans @vivek-cumulus Don't merge this yet.

Reply:
@vivek-cumulus I checked the code and it was fine. Despite that, I made changes to ensure we don't trip ourselves in the future.

@srimohans @vivek-cumulus I also realized that the code supports NLRIs along with MP_REACH so I made a change to check the validity of the nexthop in the case it has been skipped (we have MP_REACH), and NLRIs are present.